yb-voyager/src/queryissue/object_collector.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright (c) YugabyteDB, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package queryissue
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/samber/lo"
+	log "github.com/sirupsen/logrus"
+	"github.com/yugabyte/yb-voyager/yb-voyager/src/queryparser"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// ObjectCollector collects unique schema-qualified object names.
+type ObjectCollector struct {
+	objectSet map[string]bool
+}
+
+func NewObjectCollector() *ObjectCollector {
+	return &ObjectCollector{
+		objectSet: make(map[string]bool),
+	}
+}
+
+/*
+Collect processes a node and extracts object names based on node type (ignore standard SQL functions like count(), sum() etc).
+Cases covered:
+1. SELECT queries - schema name can be with table/functions
+2. Insert/Update/Delete queries
+3. TODO: cover DDLs
+
+Since Collect() will be called for each node so just the collection at concerned leaf node is enough
+*/
+func (c *ObjectCollector) Collect(msg protoreflect.Message) {
+	if msg == nil || !msg.IsValid() {
+		return
+	}
+
+	nodeType := queryparser.GetMsgFullName(msg)
+	switch nodeType {
+	// Extract table or view names in FROM clauses
+	case queryparser.PG_QUERY_RANGEVAR_NODE:
+		schemaName := queryparser.GetStringField(msg, "schemaname")
+		relName := queryparser.GetStringField(msg, "relname")
+		objectName := relName
+		if schemaName != "" {
+			objectName = fmt.Sprintf("%s.%s", schemaName, relName)
+		}
+		log.Debugf("[RangeVar] fetched schemaname=%s relname=%s objectname=%s field\n", schemaName, relName, objectName)
+		c.addObject(objectName)
+
+	// Extract target table names from DML statements
+	case queryparser.PG_QUERY_INSERTSTMT_NODE, queryparser.PG_QUERY_UPDATESTMT_NODE, queryparser.PG_QUERY_DELETESTMT_NODE:
```

Comment:
SELECT statements are already covered.
In case of SELECTs table name can only be in the FROM clause. (handled switch case 1)
Function names can be target list or from clause. (handled switch case 3)

COPY stmts are not handled right now.
I mostly focussed on the object collection from DMLs which we currently supporting in uqc

Reply:
Why we don't support COPY its also DML, and `COPY (<SELECT stmt>) TO` can have select stmt with functions objects and all?

Reply:
Ok the case is - COPY statements having SELECT inside it.
I think that will be covered, since objectCollector will be executed for each node.

@priyanshi-yb  can you share some examples? I can test that with my current code and also add to unit tests.

Reply:
> Ok the case is - COPY statements having SELECT inside it.

yeah that is the case I could think of right with COPY, not sure if there can be any, examples chatGPT gave - 
```
COPY (SELECT id, xmlagg(xmlparse(document xml_column)) AS combined_xml
      FROM my_table
      GROUP BY id) 
TO '/path/to/file.csv' WITH CSV;

COPY (SELECT ctid, xmin, id, data FROM schema_name.my_table) 
TO '/path/to/file_with_system_columns.csv' WITH CSV;
```

Reply:
Its working. Added these two cases to unit tests.
```
{
			Sql: `COPY (SELECT id, xmlagg(xmlparse(document xml_column)) AS combined_xml
					FROM my_table
						GROUP BY id)
					TO '/path/to/file.csv' WITH CSV;`,
			ExpectedObjects: []string{"xmlagg", "my_table"},
			ExpectedSchemas: []string{""},
		},
		{
			Sql: `COPY (SELECT ctid, xmin, id, data FROM schema_name.my_table) 
							TO '/path/to/file_with_system_columns.csv' WITH CSV;`,
			ExpectedObjects: []string{"schema_name.my_table"},
			ExpectedSchemas: []string{"schema_name"},
		},
```